[workflow] Fast workflow indexing

suquark opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

This PR enables indexing for workflow status. So it would be much faster to list workflows and status.

The indexing is done by creating keys under corresponding status directories. For example, `RUNNING` directory contains all keys (named with workflow ids), which the corresponding workflow is running.

One issue is that the cluster / workflow maybe crashed while updating the status, this would result in inconsistency status, because we have to create the new key, delete the old key and update workflow metadata, these actions cannot be combined as a single atomic operation. We use a special directory marking the status updating is underway. This makes us possible to detect unfinished status updating and fixing them. (See examples in newly added tests).

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

iycheng commented on python/ray/workflow/workflow_access.py:
```diff
@@ -318,8 +315,8 @@ def get_output(self, workflow_id: str, name: Optional[str]) -> WorkflowStaticRef
         wf_store = workflow_storage.WorkflowStorage(workflow_id)
-        meta = wf_store.load_workflow_meta()
-        if meta is None:
+        meta = wf_store.load_workflow_status()
```
status =

iycheng commented on python/ray/workflow/workflow_storage.py:
```diff
@@ -621,6 +677,21 @@ def delete_workflow(self) -> None:
         if not found:
             raise WorkflowNotFoundError(self._workflow_id)
 
+    def update_workflow_status(
+        self, status: WorkflowStatus, prev_status: Optional[WorkflowStatus] = None
+    ):
```
It seems a little bit complicated to me. Not quite sure why do we need `prev_status`. We can sync offline about this.

iycheng commented on python/ray/workflow/workflow_storage.py:
```diff
@@ -86,6 +87,95 @@ def is_recoverable(self) -> bool:
         )
 
 
+class WorkflowIndexingStorage:
+    """Access and maintance the indexing of workflow status."""
+
+    def __init__(self):
```
Could you add some comments here inline about the protocol here and how the fault tolerance is done?

suquark replied:
ohhh, I forget it... let me add it

iycheng commented on python/ray/workflow/workflow_storage.py:
```diff
@@ -86,6 +87,95 @@ def is_recoverable(self) -> bool:
+    def _key_workflow_with_status(self, workflow_id: str, status: WorkflowStatus):
+        return os.path.join(WORKFLOW_STATUS_DIR, status.value, workflow_id)
+
+    def _key_workflow_status_on_change(self, workflow_id: str):
```
Give some comments about what's this key for?

iycheng commented on python/ray/workflow/workflow_storage.py:
```diff
@@ -86,6 +87,124 @@ def is_recoverable(self) -> bool:
+    """Access and maintenance the indexing of workflow status.
+
+    It runs a protocol that guarantees we can recover from any interrupted
+    status updating. This protocol is **not thread-safe** for a workflow, currently
```
How about get_status in api.py? It seems that it's accessing the storage directly.

suquark replied:
this would potentially cause some inconsistency. E.g. the dirty mark is removed by each other when updating the status at the same time.

iycheng commented on python/ray/workflow/workflow_storage.py:
```diff
@@ -86,6 +87,124 @@ def is_recoverable(self) -> bool:
+
+    Here is how the protocol works:
+
+    Loading a status of a workflow
```
maybe we don't need fixing when we are doing loading/reading. With this, we can relax the thread-safety requirement I think.

Basically when listing/loading, we just do read as we need. When writing, we'll check the things there and do the fixing.

Will this be better?

suquark replied:
don't do fixing when reading/loading seems invalidate the purpose of this feature request - indexing workflow status.  

And there is another code that updates the status in parallel:

```python
def cancel(workflow_id: str) -> None:
    try:
        workflow_manager = get_management_actor()
        ray.get(workflow_manager.cancel_workflow.remote(workflow_id))
    except ValueError:
        wf_store = workflow_storage.get_workflow_storage(workflow_id)
        wf_store.update_workflow_status(WorkflowStatus.CANCELED)
```

so even updating on writing is not safe. I can still make it work but it requires a much more complicated protocol. to me it seems we should always call the workflow management actor for status loading/updating.

iycheng commented:
Sorry for the late review and thanks for working on this!

Overall, I'm good about the protocol. My concern is that this protocol enforces readers to be single thread as well which seems not to work in the current system.

One way is to ensure make read work without fixing the half updates and we only fix this when there is a write.

suquark commented:
let me try to implement another protocol using key creation time. I think it might address the multi-threading issues.

suquark commented:
@iycheng ready for reviewing again

iycheng commented on python/ray/workflow/execution.py:
```diff
@@ -128,10 +127,15 @@ def get_output(workflow_id: str, name: Optional[str]) -> ray.ObjectRef:
 def cancel(workflow_id: str) -> None:
     try:
         workflow_manager = get_management_actor()
-        ray.get(workflow_manager.cancel_workflow.remote(workflow_id))
     except ValueError:
         wf_store = workflow_storage.get_workflow_storage(workflow_id)
-        wf_store.save_workflow_meta(WorkflowMetaData(WorkflowStatus.CANCELED))
+        # TODO(suquark): Here we update workflow status "offline", so it is likely
+        # thread-safe because there is no workflow management actor updating the
+        # workflow concurrently. But we should be careful if we are going to
+        # update more workflow status offline in the future.
+        wf_store.update_workflow_status(WorkflowStatus.CANCELED)
+        return
+    ray.get(workflow_manager.cancel_workflow.remote(workflow_id))
```
Should we move `ray.get` back to try block?

suquark replied:
no, I think we put `ray.get` in try block accidentally. `ray.get` cannot generate `ValueError`, only `get_management_actor` generates `ValueError`

iycheng commented on python/ray/workflow/workflow_storage.py:
```diff
@@ -86,6 +87,129 @@ def is_recoverable(self) -> bool:
+                )
+            self._storage.delete(self._key_workflow_status_dirty(workflow_id))
+
+    def load_workflow_status(self, workflow_id: str):
+        """Load the committed workflow status."""
+        raw_data = self._storage.get(self._key_workflow_metadata(workflow_id))
+        if raw_data is not None:
+            metadata = json.loads(raw_data)
+            return WorkflowStatus(metadata["status"])
+        return WorkflowStatus.NONE
```
So here we only return the committed status. Could it become an issue if it has a dirty flag? I remember it's handled in the application layer (api.py).

suquark replied:
No, because `committed status` is the status we will return even if the dirty flag exists.

iycheng commented on python/ray/workflow/workflow_storage.py:
```diff
@@ -86,6 +87,129 @@ def is_recoverable(self) -> bool:
+            return WorkflowStatus(metadata["status"])
+        return WorkflowStatus.NONE
+
+    def list_workflow(self) -> List[Tuple[str, WorkflowStatus]]:
```
I think we need an extra param here `list_workflow(self, status=None)`

Here if status is set, we'll only check the dirty directory and the specified status directory. I'm ok with another PR to fix this

suquark replied:
I just implemented it.

iycheng commented on python/ray/workflow/workflow_storage.py:
```diff
@@ -86,6 +87,129 @@ def is_recoverable(self) -> bool:
+                        self._storage.delete(
+                            self._key_workflow_with_status(workflow_id, s)
+                        )
+            else:
```
we should always set the dirty here, right?

suquark replied:
no, because we did not delete the existing dirty flag

iycheng replied:
Got it!

iycheng commented on python/ray/workflow/workflow_storage.py:
```diff
@@ -86,6 +87,129 @@ def is_recoverable(self) -> bool:
+                    self._key_workflow_with_status(workflow_id, prev_status), b""
+                )
+                for s in WorkflowStatus:
+                    if s != prev_status:
+                        self._storage.delete(
+                            self._key_workflow_with_status(workflow_id, s)
+                        )
```
No need to update this PR. But do you think in the future we can put the status in the dirty flag and only delete that?

suquark replied:
good idea

iycheng commented:
It seems there is a bug here: we should always set up the flag I think.

Another thing is that, I think we need to put status filter into storage layer so that we don't need to read successful workflow status which is not useful for [resume_all](https://github.com/ray-project/ray/blob/master/python/ray/workflow/api.py#L298)/[list_all](https://github.com/ray-project/ray/blob/master/python/ray/workflow/api.py#L243) with filter. I'm OK with this PR and having another one for this optimizing.

suquark commented:
@iycheng I think we did not set the dirty flag because in that branch we already detected the dirty flag. Since the workflow status updating is single threaded, there is no need to create it again. (Create it again also does not work under concurrent case, because another faster process could delete the newly created flag anyway - the order of create/delete of different processes could be arbitrary).

iycheng commented:
Got it and thanks for the explaination.

suquark commented:
@iycheng I just updated the PR and support status filter. I also fixed a bug: in the original `workflow.api.list_all()`, `        status_filter = set(WorkflowStatus.__members__.keys())` returns a set of strings instead of a set of workflow status. It is clearly against the behaviors of other branches in the function, and this bug reveals when I implement status filter in the storage layer.

iycheng commented on python/ray/workflow/workflow_storage.py:
```diff
@@ -167,27 +167,38 @@ def load_workflow_status(self, workflow_id: str):
+        """
+        if status_filter is None:
+            status_filter = set(WorkflowStatus)
+        status_filter.discard(WorkflowStatus.NONE)
```
We should either assert None is not there or print a warning.

iycheng reviewed (approved):
LGTM!

suquark commented:
CI failures seem unrelated. I'll merge this PR.
